### PR TITLE
[proposal] New unit test pattern

### PR DESCRIPTION
**Pull request by ruanxin**

### CODE REVIEW

File: operator/controllers/kyma_test.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	AfterAll(func() {
+		Expect(controlPlaneClient.Delete(ctx, kyma)).Should(Succeed())
+	})
+	DescribeTable("Testing ModuleInfo",
+		func(condition func() bool, expectedBehavior func() bool) {
+			Eventually(condition, timeout*2, interval).Should(BeTrue())
+			Eventually(expectedBehavior, timeout*2, interval).Should(BeTrue())
+		},
+		Entry("When init state, expect number of ModuleInfo matches spec.modules", noCondition(), expectedCorrectNumberOfModuleInfo(kyma.Name)),
+		Entry("When remove module in spec, expect number of ModuleInfo matches spec.modules", removeModule(kyma.Name), expectedCorrectNumberOfModuleInfo(kyma.Name)),
```
Comment (ruanxin) [inline operator/controllers/kyma_test.go:65]: This test case currently failed, and it's true positive bug...

File: operator/controllers/kyma_controller_helper_test.go [hunk 2]
```diff
@@ -37,6 +40,37 @@ func NewTestKyma(name string) *v1alpha1.Kyma {
...
+}
+
+const letterBytes = "abcdefghijklmnopqrstuvwxyz"
+const (
+	letterIdxBits = 6                    // 6 bits to represent a letter index
+	letterIdxMask = 1<<letterIdxBits - 1 // All 1-bits, as many as letterIdxBits
+	letterIdxMax  = 63 / letterIdxBits   // # of letter indices fitting in 63 bits
+)
+
+func RandString(n int) string {
```
Comment (jakobmoellersap) [inline operator/controllers/kyma_controller_helper_test.go:54]: cool stack overflow copy :D

File: operator/controllers/kyma_test.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+		Name:           NewUniqModuleName(),
+		Channel:        v1alpha1.ChannelStable,
+	})
+
+	moduleTemplates := make([]*v1alpha1.ModuleTemplate, 0)
+
+	BeforeAll(func() {
+		deployModuleTemplate(kyma, moduleTemplates)
+	})
+	AfterAll(func() {
```
Comment (jakobmoellersap) [inline operator/controllers/kyma_test.go:56]: this I would put under NewTestKymaWithCleanup

**Review by jakobmoellersap — COMMENTED** [review]
Great stuff! I think we should structure all tests like this

File: operator/controllers/kyma_controller_helper_test.go (hunk 2 above)
Comment (ruanxin) [reply operator/controllers/kyma_controller_helper_test.go:48]: use a simplified version

File: operator/controllers/kyma_test.go (hunk 3 above)
Comment (ruanxin) [reply operator/controllers/kyma_test.go]: what do you mean? extract it to be a function?

**Review by jakobmoellersap — DISMISSED** [review]
💯 very cool table driven testing! you should show this on review. anyhow lint is still failing, probably need to reapprove later

**Review by jakobmoellersap — APPROVED** [review]

**Pull request merged**